sky/server/requests/requests.py

```diff
@@ -572,36 +592,86 @@ def reset_db_and_logs():
                   ignore_errors=True)
 
 
-def request_lock_path(request_id: str) -> str:
-    lock_path = os.path.expanduser(REQUEST_LOG_PATH_PREFIX)
-    os.makedirs(lock_path, exist_ok=True)
-    return os.path.join(lock_path, f'.{request_id}.lock')
+def _update_request(request: Request, fields: Optional[List[str]] = None):
+    """Update a REST request in the database."""
+    if fields is None:
+        fields = REQUEST_COLUMNS
+    assert _DB is not None
+    sql_statement = (f'UPDATE {REQUEST_TABLE} SET '
+                     f'{", ".join([f"{col} = ?" for col in fields])} '
+                     f'WHERE request_id = ?')
+    with _DB.conn:
+        cursor = _DB.conn.cursor()
+        cursor.execute(sql_statement,
+                       request.to_row(fields) + (request.request_id,))
 
 
-@contextlib.contextmanager
 @init_db
 @metrics_lib.time_me
-def update_request(request_id: str) -> Generator[Optional[Request], None, None]:
-    """Get and update a SkyPilot API request."""
-    # Acquire the lock to avoid race conditions between multiple request
-    # operations, e.g. execute and cancel.
-    with filelock.FileLock(request_lock_path(request_id)):
-        request = _get_request_no_lock(request_id)
-        yield request
-        if request is not None:
-            _add_or_update_request_no_lock(request)
+def update_request(
+    request_id: str,
+    set_status: Optional[RequestStatus] = None,
+    set_should_retry: Optional[bool] = None,
+    set_pid: Optional[int] = None,
+    unset_pid: bool = False,
+    set_finished_at: Optional[float] = None,
+    set_error: Optional[BaseException] = None,
+    match_status: Optional[List[RequestStatus]] = None,
+) -> Optional[Request]:
+    """Update a SkyPilot API request in the database."""
+    assert _DB is not None
 
+    update_params = []
+    update_statement = f'UPDATE {REQUEST_TABLE} SET '
+    filter_statements = []
+    if set_status is not None:
+        filter_statements.append('status = ?')
+        update_params.append(set_status.value)
+    if set_should_retry is not None:
+        filter_statements.append('should_retry = ?')
+        update_params.append(set_should_retry)
+    if set_pid is not None:
+        filter_statements.append('pid = ?')
+        update_params.append(set_pid)
+    if unset_pid:
+        filter_statements.append('pid = NULL')
+    if set_finished_at is not None:
+        filter_statements.append('finished_at = ?')
+        update_params.append(set_finished_at)
+    if set_error is not None:
+        filter_statements.append('error = ?')
+        update_params.append(json.dumps(_serialize_error(set_error)))
+    update_statement += f'{", ".join(filter_statements)} '
+    update_statement += 'WHERE request_id = ? '
+    update_params.append(request_id)
+    if match_status is not None:
+        update_statement += (
+            'AND status IN '
+            f'({", ".join([repr(status.value) for status in match_status])}) ')
```

Comment:
Ideally we use parameters for these values as well. Would like all the string interpolation to be constants.